Feature: Edit note UI

Roshan-y opened the pull request:
# Description

Updated edit note ui. Made multiple screens to declutter the edit note screen. We have a general, comments, pdf and content screen. Here is a video displaying the new ui:
[Screen Recording 2024-11-22 at 02.18.13.zip](https://github.com/user-attachments/files/17853931/Screen.Recording.2024-11-22.at.02.18.13.zip)

I have created a common delete popup composable that everyone can use in ui.common.DeletePopup.kt. I have additionally created a common loading screen.

# How Has This Been Tested?

### UI Tests
I have updated the existing tests and added UI tests for each screen. I have properly mocked the necessary parameters to be able to fully test each screen.

### End to end Tests
I have corrected the unit tests to work with the new screens

# Dependencies

- `com.google.accompanist:accompanist-flowlayout` to be able to use FlowRow. Here is further [documentation](https://developer.android.com/develop/ui/compose/layouts/flow)
- `io.github.afreakyelf:Pdf-Viewer` to be able to display the pdf in the pdf view screen. Here is further [documentation](https://github.com/afreakyelf/Pdf-Viewer)


# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged in downstream modules, no conflicts with main

MatthieuSLR9 reviewed (commented):
See PR review comment bellow.

MatthieuSLR9 reviewed (changes requested):
Excellent features ! The code is very well documented. Some minor changes could be done however. Furthermore on testing the features I noticed that there is a bug where you can not deselect a text when you started to input into it. This problem is minor and could be fixed in another PR.

Roshan-y commented on app/src/main/java/com/github/onlynotesswent/ui/overview/editnote/EditNoteNavigationMenu.kt:
```diff
@@ -0,0 +1,107 @@
+                navigateTo?.let { navigationActions.navigateTo(it) }
+                onClick()
+              }) {
+                Text(text = "Discard", color = MaterialTheme.colorScheme.error)
```
I updated the common DeletePopup to a ConfirmationPopup that we can use for all popups to make  it uniform

Roshan-y commented on app/src/main/java/com/github/onlynotesswent/ui/overview/editnote/Comments.kt:
```diff
@@ -0,0 +1,206 @@
+                      updatedComments.editCommentByCommentId(comment.commentId, it)
+                  onCommentsChange(updatedCollection)
+                },
+                label = {
```
When testing if the text is too long it moves the text in the line below to show the whole comment, so I do not think it is necessary.

MatthieuSLR9 replied:
Yes I know as I wrote this part of the code but I meant that if someone wrote thousands of lines of comments it would become unusable. As there is no way to close a comment the user would have to scroll a lot. However this problem can be solved in another PR. I will implement in the future the option to close a comment to fix this issue.

Roshan-y replied:
Ah yes my bad I misread your comment

Roshan-y commented on app/src/main/java/com/github/onlynotesswent/ui/overview/editnote/Comments.kt:
```diff
@@ -0,0 +1,206 @@
+                value = comment.content,
+                onValueChange = {
+                  val updatedCollection =
+                      updatedComments.editCommentByCommentId(comment.commentId, it)
```
Yes valid point. Instead of modifying it here I modified line 69-72 and 80-83 so that the empty comments don't get saved in the database:
` val commentsNotEmpty = updatedComments.commentsList.filter { it.content.isNotEmpty() }
              noteViewModel.updateNote(
                  note!!.copy(comments = Note.CommentCollection(commentsNotEmpty)),
                  currentUser!!.uid)`

Roshan-y commented on app/src/main/java/com/github/onlynotesswent/ui/overview/editnote/EditMarkdownScreen.kt:
```diff
@@ -276,7 +307,7 @@ fun ControlWrapper(
               modifier = Modifier.size(20.dp).testTag("FilterChipIcon"),
               imageVector = Icons.Default.Check,
-              contentDescription = if (selected) "Selected" else "Not Selected",
+              contentDescription = "Selected",
```
Not necessary anymore as in line 306 I first check if it is selected.

MatthieuSLR9 replied:
Yes I see but I would have changed the logic to display a not Selected content description when the icon is not selected.

Roshan-y commented on app/src/main/java/com/github/onlynotesswent/ui/overview/editnote/EditNote.kt:
```diff
@@ -0,0 +1,487 @@
+                comments = note.comments),
+            currentUser.uid)
+        noteViewModel.getNoteById(note.id)
+        Toast.makeText(context, "Note saved", Toast.LENGTH_SHORT).show()
```
Yes agree, but I think all toasts should be changed to snack bar so I will leave this for a future pr.

Roshan-y commented on app/src/main/java/com/github/onlynotesswent/ui/common/DeletePopup.kt:
```diff
@@ -0,0 +1,59 @@
+ * @param onDismiss The action to perform when the user dismisses the dialog.
+ */
+@Composable
+fun DeletePopup(title: String, text: String, onConfirm: () -> Unit, onDismiss: () -> Unit) {
```
See later comment as I modify Delete popup to confirmation popup

Roshan-y commented:
> Excellent features ! The code is very well documented. Some minor changes could be done however. Furthermore on testing the features I noticed that there is a bug where you can not deselect a text when you started to input into it. This problem is minor and could be fixed in another PR.

I have changed what you have requested. The points I thought that weren't necessary I commented my reasoning below the comments. I also modified the detail page of the edit note to be simpler as per the assistants request, I have attached a video but I think you should test it out too:


https://github.com/user-attachments/assets/3c4f3c8c-95dc-471b-bd7d-1ffefc6735cd
